entity-registry/src/main/java/com/linkedin/metadata/aspect/hooks/FieldPathMutator.java

```diff
@@ -0,0 +1,142 @@
+import com.linkedin.metadata.aspect.ReadItem;
+import com.linkedin.metadata.aspect.RetrieverContext;
+import com.linkedin.metadata.aspect.batch.BatchItem;
+import com.linkedin.metadata.aspect.batch.ChangeMCP;
+import com.linkedin.metadata.aspect.plugins.config.AspectPluginConfig;
+import com.linkedin.metadata.aspect.plugins.hooks.MutationHook;
+import com.linkedin.schema.EditableSchemaFieldInfo;
+import com.linkedin.schema.EditableSchemaFieldInfoArray;
+import com.linkedin.schema.EditableSchemaMetadata;
+import com.linkedin.schema.SchemaField;
+import com.linkedin.schema.SchemaFieldArray;
+import com.linkedin.schema.SchemaMetadata;
+import com.linkedin.util.Pair;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Setter
+@Getter
+@Accessors(chain = true)
+public class FieldPathMutator extends MutationHook {
+  @Nonnull private AspectPluginConfig config;
+
+  @Override
+  protected Stream<Pair<ChangeMCP, Boolean>> writeMutation(
+      @Nonnull Collection<ChangeMCP> changeMCPS, @Nonnull RetrieverContext retrieverContext) {
+
+    List<Pair<ChangeMCP, Boolean>> results = new LinkedList<>();
+
+    for (ChangeMCP item : changeMCPS) {
+      if (changeTypeFilter(item) && aspectFilter(item)) {
+        if (item.getAspectName().equals(SCHEMA_METADATA_ASPECT_NAME)) {
+          results.add(Pair.of(item, processSchemaMetadataAspect(item)));
+        } else {
+          results.add(Pair.of(item, processEditableSchemaMetadataAspect(item)));
+        }
+      } else {
+        // no op
+        results.add(Pair.of(item, false));
+      }
+    }
+
+    return results.stream();
+  }
+
+  /*
+    TODO: After some time, this should no longer be required. Assuming at least 1 write has
+          occurred for all schema aspects.
+  */
+  @Override
+  protected Stream<Pair<ReadItem, Boolean>> readMutation(
+      @Nonnull Collection<ReadItem> items, @Nonnull RetrieverContext retrieverContext) {
+    List<Pair<ReadItem, Boolean>> results = new LinkedList<>();
+
+    for (ReadItem item : items) {
+      if (aspectFilter(item)) {
+        if (item.getAspectName().equals(SCHEMA_METADATA_ASPECT_NAME)) {
+          results.add(Pair.of(item, processSchemaMetadataAspect(item)));
+        } else {
+          results.add(Pair.of(item, processEditableSchemaMetadataAspect(item)));
+        }
+      } else {
+        // no op
+        results.add(Pair.of(item, false));
+      }
+    }
+
+    return results.stream();
+  }
+
+  private static boolean changeTypeFilter(BatchItem item) {
+    return !ChangeType.DELETE.equals(item.getChangeType())
+        && !ChangeType.PATCH.equals(item.getChangeType());
```

Comment:
I believe the idea here is that it can't occur for patches because patches require a keyed operation which prevents duplicates, but @david-leifker please verify.

Reply:
PATCH is not needed since mutation occurs after PATCH change type has been converted to UPSERT